examples/linkedlist.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import print_function, absolute_import
+from numba.utils import OrderedDict
+import numpy as np
+from numba import njit
+from numba.jitclass import jitclass
+from numba import int32, deferred_type, optional
+from numba.runtime import rtsys
+
+node_type = deferred_type()
+
+spec = OrderedDict()
+spec['data'] = int32
+spec['next'] = optional(node_type)
+
+
+@jitclass(spec)
+class LinkedNode(object):
+    def __init__(self, data, next):
+        self.data = data
+        self.next = next
+
+    def prepend(self, data):
+        return LinkedNode(data, self)
+
+
+@njit
+def make_linked_node(data):
+    return LinkedNode(data, None)
+
+
+node_type.define(LinkedNode.class_type.instance_type)
```

Comment:
> Can you elaborate on "leaking in all aspects of typing and code generation". Are you referring to the cyclic reference?

No, I mean handling the deferred type adds some special cases at various points along the compilation pipeline (because the deferred type is semantically equivalent to the "real" type, but the type system doesn't allow recognizing that automatically -- and I have no idea how we could change that).

> 2) I don't understand how node_type = node_type.resolve(LinkedNode.class_type.instance_type) would work. 

Yes, sorry, my suggestion was wrong. My bad.

> Does deferred('LinkedNode') require registering the name "LinkedNode" into the type system?

No, I think it could work simply as a local name for resolution when a jitclass is compiled. Perhaps I'm missing something, though.
I agree registering the name into the type system would add a layer of complexity that's probably not welcome.